File: Wdtc/src/main/java/org/wdt/wdtc/ui/JavaFxUtils.java
package org.wdt.wdtc.ui;

import com.google.gson.JsonArray;
import org.apache.log4j.Logger;
import org.wdt.utils.dependency.DependencyDownload;
import org.wdt.utils.gson.Json;
import org.wdt.utils.gson.JsonArrayUtils;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.utils.io.FileUtils;
import org.wdt.utils.io.FilenameUtils;
import org.wdt.utils.io.IOUtils;
import org.wdt.wdtc.core.download.SpeedOfProgress;
import org.wdt.wdtc.core.game.LibraryObject;
import org.wdt.wdtc.core.manger.FileManger;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.ThreadUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.File;
import java.io.IOException;
import java.lang.module.Configuration;
import java.lang.module.ModuleFinder;
import java.lang.module.ModuleReference;
import java.net.URL;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public class JavaFxUtils {
    private static final Logger logmaker = WdtcLogger.getLogger(JavaFxUtils.class);


    public static void setJavaFXListJson() {
        try {
            File OpenJfxListFile = new File(FileManger.getWdtcImplementationPath(), "openjfx-list.json");
            if (FileUtils.isFileNotExistsAndIsNotSameSize(OpenJfxListFile, 2393)) {
                List<String> MoudleList = List.of("javafx.base", "javafx.controls", "javafx.fxml", "javafx.web", "javafx.graphics", "javafx.media");
                JsonArray array = new JsonArray();
                for (String s : MoudleList) {
                    String path = "org/openjfx/" + s.replaceAll("\\.", "-") + "/17.0.6/" + s.replaceAll("\\.", "-") + "-17.0.6-win.jar";
                    URL url = new URL("https://maven.aliyun.com/repository/public/" + path);
                    LibraryObject.Artifact artifact = new LibraryObject.Artifact();
                    artifact.setUrl(url);
                    artifact.setPath(FilenameUtils.separatorsToUnix(path));
                    artifact.setSha1(IOUtils.getInputStreamSha1(url.openStream()));
                    artifact.setSize(url.openConnection().getContentLength());
                    LibraryObject.Downloads downloads = new LibraryObject.Downloads();
                    downloads.setArtifact(artifact);
                    LibraryObject libraryObject = new LibraryObject();
                    libraryObject.setDownloads(downloads);
                    libraryObject.setLibraryName("org.openjfx:" + s + ":win:17.0.6");
                    array.add(Json.GSON.toJsonTree(libraryObject, LibraryObject.class));
                }
                JsonUtils.writeObjectToFile(OpenJfxListFile, array);
            }
        } catch (IOException e) {
            logmaker.warn(WdtcLogger.getExceptionMessage(e));
        }
    }

    public static void downloadDependencies() throws IOException {
        JsonArray array = getArrayObject();
        SpeedOfProgress speed = new SpeedOfProgress(array.size());
        for (int i = 0; i < array.size(); i++) {
            LibraryObject libraryObject = LibraryObject.getLibraryObject(array.get(i).getAsJsonObject());
            LibraryObject.Artifact artifact = libraryObject.getDownloads().getArtifact();
            File Library = new File(FileManger.getWtdcOpenJFXPath(), artifact.getPath());
            ThreadUtils.startThread(() -> {
                try {
                    if (FileUtils.isFileNotExistsAndIsNotSameSize(Library, artifact.getSize())) {
                        DownloadUtils.startDownloadTask(artifact.getUrl(), Library);
                    }
                    speed.countDown();
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
        }
        speed.await();
    }

    private static JsonArray getArrayObject() throws IOException {
        return JsonArrayUtils.parseJsonArray(IOUtils.toString(Objects.requireNonNull(JavaFxUtils.class.getResourceAsStream("/openjfx-list.json"))));
    }

    public static void loadJavaFXPatch() {
        try {
            Class.forName("javafx.application.Application");
        } catch (ClassNotFoundException e) {
            try {
                logmaker.info("Load JavaFX Dependencies");
                JsonArray array = getArrayObject();
                Set<Path> jarPaths = new HashSet<>();
                Set<String> modules = new HashSet<>();
                for (int i = 0; i < array.size(); i++) {
                    LibraryObject libraryObject = LibraryObject.getLibraryObject(array.get(i).getAsJsonObject());
                    LibraryObject.Artifact artifact = libraryObject.getDownloads().getArtifact();
                    File library = new File(FileManger.getWtdcOpenJFXPath(), artifact.getPath());
                    jarPaths.add(library.toPath());
                    modules.add(new DependencyDownload(libraryObject.getLibraryName()).getArtifactId());
                }
                // Form : HMCL3
                ModuleFinder finder = ModuleFinder.of(jarPaths.toArray(Path[]::new));
                for (ModuleReference mref : finder.findAll()) {
                    ((jdk.internal.loader.BuiltinClassLoader) ClassLoader.getSystemClassLoader()).loadModule(mref);
                }
                Configuration config = Configuration.resolveAndBind(finder, List.of(ModuleLayer.boot().configuration()), finder, modules);
                ModuleLayer.defineModules(config, List.of(ModuleLayer.boot()), name -> ClassLoader.getSystemClassLoader());
                logmaker.info("Done");
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        }
    }

    public static void ckeckJavaFX() throws IOException {
        downloadDependencies();
        loadJavaFXPatch();
    }
}

File: Wdtc/src/main/kotlin/org/wdt/wdtc/ui/JavaFxUtils.kt
@file:JvmName("JavaFXUtils")

package org.wdt.wdtc.ui

import jdk.internal.loader.BuiltinClassLoader
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.wdt.utils.gson.getJsonArray
import org.wdt.utils.gson.parseArray
import org.wdt.utils.gson.parseJsonObject
import org.wdt.utils.gson.writeObjectToFile
import org.wdt.utils.io.IOUtils
import org.wdt.utils.io.getSha1
import org.wdt.wdtc.core.download.SpeedOfProgress
import org.wdt.wdtc.core.game.LibraryObject
import org.wdt.wdtc.core.game.LibraryObjectList
import org.wdt.wdtc.core.game.serializeLibraryObjectListGsonBuilder
import org.wdt.wdtc.core.game.toDependency
import org.wdt.wdtc.core.manger.SystemKind.MAC
import org.wdt.wdtc.core.manger.SystemKind.WINDOWS
import org.wdt.wdtc.core.manger.currentSystem
import org.wdt.wdtc.core.manger.wdtcDependenciesDirectory
import org.wdt.wdtc.core.manger.wtdcOpenJFXDirectory
import org.wdt.wdtc.core.utils.*
import org.wdt.wdtc.ui.window.setErrorWin
import java.io.File
import java.io.IOException
import java.lang.module.Configuration
import java.lang.module.ModuleFinder
import java.net.URL
import java.nio.file.Path

val openJFXListFile = File(wdtcDependenciesDirectory, "openjfx-list.json")
val osName: String
  get() {
    if (currentSystem == MAC) error("Wdtc cannot run on macos!")
    return if (currentSystem == WINDOWS) "win" else "linux"
  }
val canRunOSList = listOf("win", "linux")

fun setJavaFXListJson() {
  try {
    val moudleList =
      listOf("javafx.base", "javafx.controls", "javafx.fxml", "javafx.web", "javafx.graphics", "javafx.media")
    val runtimeListMap = HashMap<String, LibraryObjectList>()
    canRunOSList.forEach { os ->
      val runtimeList = LibraryObjectList().apply {
        moudleList.forEach { moudleName ->
          val path = moudleName.replace(".", "-").let {
            "org/openjfx/$it/17.0.6/$it-17.0.6-$os.jar"
          }
          val artifact = URL("https://maven.aliyun.com/repository/public/$path").let {
            LibraryObject.Artifact(
              url = it,
              path = path,
              sha1 = it.openStream().getSha1(),
              size = it.openConnection().contentLength.toLong(),
            )
          }
          LibraryObject(
            downloads = LibraryObject.Downloads(
              artifact = artifact
            ), libraryName = "org.openjfx:$moudleName:win:17.0.6".toDependency()
          ).addToList()

        }
      }
      runtimeListMap[os] = runtimeList
    }
    openJFXListFile.writeObjectToFile(runtimeListMap, serializeLibraryObjectListGsonBuilder)
  } catch (e: IOException) {
    logmaker.warning(e.getExceptionMessage())
  }
}

fun downloadDependencies() = runBlocking {
  runtimeList.run {
    val speed = SpeedOfProgress(size).apply {
      coroutineScope = executorCoroutineScope(size, "Download javafx jar file")
    }
    forEach {
      val artifact = it.downloads.artifact
      val library = File(wtdcOpenJFXDirectory, artifact.path)
      speed.coroutineScope.launch {
        try {
          if (library.compareFile(artifact)) {
            startDownloadTask(artifact.url, library)
            speed.countDown()
          } else {
            speed.countDown()
          }
        } catch (e: IOException) {
          setErrorWin(e)
        }
      }
    }
    speed.await()
  }
}

private val runtimeList: LibraryObjectList
  get() = IOUtils.toString(getResourceAsStream("/openjfx-list.json")).parseJsonObject().getJsonArray(osName)
    .parseArray(serializeLibraryObjectListGsonBuilder)

fun loadJavaFXPatch() {
  try {
    Class.forName("javafx.application.Application")
  } catch (e: ClassNotFoundException) {
    try {
      logmaker.info("Load JavaFX Dependencies")
      logmaker.info("OpenJFX Platform:$osName")
      val jarPaths: MutableSet<Path> = HashSet()
      val modules: MutableSet<String> = HashSet()
      runtimeList.forEach {
        it.run {
          File(wtdcOpenJFXDirectory, downloads.artifact.path).run {
            jarPaths.add(toPath())
          }
          modules.add(libraryName.artifactId)
        }
      }
      // Form : HMCL3
      val finder = ModuleFinder.of(*jarPaths.toTypedArray())
      for (mref in finder.findAll()) {
        (ClassLoader.getSystemClassLoader() as BuiltinClassLoader).loadModule(mref)
      }
      val config = Configuration.resolveAndBind(finder, listOf(ModuleLayer.boot().configuration()), finder, modules)
      ModuleLayer.defineModules(
        config, listOf(ModuleLayer.boot())
      ) { ClassLoader.getSystemClassLoader() }
      logmaker.info("Done")
    } catch (ex: IOException) {
      setErrorWin(ex)
    }
  }
}

fun ckeckJavaFX() {
  downloadDependencies()
  loadJavaFXPatch()
}

